src/ert/ensemble_evaluator/evaluator_async.py

```diff
@@ -0,0 +1,474 @@
+        # * https://github.com/equinor/ert/issues/1250
+        self._iter: int = iter_
+        self._config: EvaluatorServerConfig = config
+        self._ensemble: Ensemble = ensemble
+
+        self._loop: Optional[asyncio.AbstractEventLoop] = None
+
+        self._clients: Set[WebSocketServerProtocol] = set()
+        self._dispatchers_connected: asyncio.Queue[None] = asyncio.Queue()
+
+        self._events: asyncio.Queue[CloudEvent] = asyncio.Queue()
+        self._messages_to_send: asyncio.Queue[str] = asyncio.Queue()
+
+        self._result = None
+
+        self._ee_tasks: List[asyncio.Task[None]] = []
+        self._server_started: asyncio.Event = asyncio.Event()
+        self._server_done: asyncio.Event = asyncio.Event()
+
+        # batching section
+        self._batch_processing_queue: asyncio.Queue[
+            List[Tuple[Callable[[List[CloudEvent]], Awaitable[None]], CloudEvent]]
+        ] = asyncio.Queue()
+        self._max_batch_size: int = 500
+        self._batching_interval: int = 2
+
+    async def _publisher(self) -> None:
+        while True:
+            msg = await self._messages_to_send.get()
+            await asyncio.gather(
+                *[client.send(msg) for client in self._clients],
+                return_exceptions=True,
+            )
+            self._messages_to_send.task_done()
+
+    async def _append_message(self, snapshot_update_event: PartialSnapshot) -> None:
+        message = self._create_cloud_message(
+            EVTYPE_EE_SNAPSHOT_UPDATE,
+            snapshot_update_event.to_dict(),
+        )
+        if message:
+            await self._messages_to_send.put(message)
+
+    async def _process_event_buffer(self) -> None:
+        while True:
+            batch = await self._batch_processing_queue.get()
+            function_to_events_map: Dict[
+                Callable[[List[CloudEvent]], Awaitable[None]], List[CloudEvent]
+            ] = {}
+            for func, event in batch:
+                if func not in function_to_events_map:
+                    function_to_events_map[func] = []
+                function_to_events_map[func].append(event)
+
+            for func, events in function_to_events_map.items():
+                await func(events)
+
+            self._batch_processing_queue.task_done()
+
+    async def _batch_events_into_buffer(self) -> None:
+        event_handler = {}
+
+        def set_event_handler(event_types: Set[str], function: Any) -> None:
+            for event_type in event_types:
+                event_handler[event_type] = function
+
+        for e_type, f in (
+            (EVGROUP_FM_ALL, self._fm_handler),
+            ({EVTYPE_ENSEMBLE_STARTED}, self._started_handler),
+            ({EVTYPE_ENSEMBLE_STOPPED}, self._stopped_handler),
+            ({EVTYPE_ENSEMBLE_CANCELLED}, self._cancelled_handler),
+            ({EVTYPE_ENSEMBLE_FAILED}, self._failed_handler),
+        ):
+            set_event_handler(e_type, f)
```

Comment:
Instead of having this as a loop and manually creating the set of item; can we call the function on the items one by one instead?

Reply:
But this is not a biggy, since this runs only once

Reply:
I think it would be more readable to hard code it directly
```
            set_event_handler(EVGROUP_FM_ALL, self._fm_handler)
            set_event_handler({EVTYPE_ENSEMBLE_STARTED}, self._started_handler)
            set_event_handler({EVTYPE_ENSEMBLE_STOPPED}, self._stopped_handler)
            set_event_handler({EVTYPE_ENSEMBLE_CANCELLED}, self._cancelled_handler)
            set_event_handler({EVTYPE_ENSEMBLE_FAILED}, self._failed_handler)
```

Reply:
You might be right there :+1: